lib/spack/spack/build_systems/autotools.py

```diff
@@ -259,7 +272,8 @@ def autoreconf(self, spec, prefix):
             return
         # Else try to regenerate it
         autotools = ['m4', 'autoconf', 'automake', 'libtool']
-        missing = [x for x in autotools if x not in spec]
+        # Previously, spec had only dependencies. Now they are in spec._dependencies
+        missing = [x for x in autotools if x not in spec._dependencies]
```

Comment:
You should use the public api and limit to direct build type dependencies: `spec.dependencies(deptype='build')` and then never use `in`:

```suggestion
        build_deps = [d.name for d in s.dependencies(deptype='build')]
        missing = [x for x in autotools if x not in build_deps]
```

Reply:
Would also be good to add a test for this...

Reply:
Looks great, much better! With that, I get `NameError: name 's' is not defined`

I assume you mean `spec.dependencies`, like this:
```py
build_deps = [d.name for d in spec.dependencies(deptype='build')]
```
Another improvement proposal: 

As autoconf has m4 as run-dependency, m4 does not have to be added as `depends_on()` explicitly. `depends_on(autoconf)1 implies that m4 (indirectly) is provided as well, so packages do not have to add it explicitly to each package:
```diff
-        autotools = ['m4', 'autoconf', 'automake', 'libtool']
+        autotools = ['autoconf', 'automake', 'libtool']
```
This removes the need of having this in all master-building packages:
```py
+    depends_on('m4', when='@master', type='build')
```
Just this is enough (tested)
```py
+    depends_on('autoconf', when='@master', type='build')
+    depends_on('automake', when='@master', type='build')
+    depends_on('libtool',  when='@master', type='build')
```
I'll push that change to PR #26101 first, then continue here when only the check in #26101 is approved and fully tested in all the tests (and yes, we'll likely want/need a test for this check too). Any hints and help would be welcome.